### PR TITLE
Feature/cli dependency validation

### PR DESCRIPTION
**Pull request by hgpestana**
## What does this do
This PR gives *einstein* the possibility of validating local dependencies before submitting a development to a git repository with a tag. Previously, in order to validate the dependencies, the user would need to submit the code and create a tag and, only then, could he or she validate the *einstein* dependencies.  

## How does this work?
The already provided functionality from *einstein* was split. Two new commands where added, a _validate_ and a _calculate_ command. The calculate command provides the already existing functionality, with the added bonus that now you can also log the output to a folder/file of your choosing and also have a more granular control over the verbosity. The validate command is very similar, with the difference that instead of providing a project, you'll need to provide an *einstein.yaml* input file to be validated.

Examples:

```bash
einstein validate -i=c:\dev\project\einstein.yaml -vvv -lt logs.txt -o dependencies.json
```
```bash
einstein calculate -p=namespace/project:1.0.0 -vvv -lt logs.txt -o dependencies.json
```

For both commands you can call the help commands for further information about the supported functionality:

```bash
einstein validate help
```
```bash
einstein calculate help
```

## How can this be tested?
Just run the _validate_ or _calculate_ command options against an existing *einstein.yaml* or a project.

### CODE REVIEW

**Comment by crisgoncalves** [comment]
Hey @hgpestana.

Regarding the `validate` command, I got some unexpected results.
By trying  to validate `ebanka-venus` dependencies, for its `snapshot` version, I got the following result:

` {
    "canais-n-presenciais/backoffice": "v5.7.0",
    "middleware/acm-mediator-service-parent": "1.0.1",
    "canais-n-presenciais/filemanager": "v2.0.0",
    "middleware/filemanager-jasper-reports": "v1.2.0",
    "middleware/ebankaplus-jasper-reports": "v1.1.0",
    "middleware/irc-custom-bcga-ao-parent": "2.1.1",
    "middleware/cpi-invoice-service-parent": "1.0.1",
    "middleware/psd2-integration-layer-parent": "3.1.0-SNAPSHOT"
}
`  

According to project's `einstein.yaml` file, they're missing dependencies...

On the other hand, I noticed that 'calculated dependencies per project' comes in a strange format:   

`[canais-n-presenciais/backoffice:[5.7.0:], middleware/irc-ws-bb-parent:[4.0.0:, 4.3.3:middleware/psd2-integration-layer-parent:3.1.0-SNAPSHOT, 4.4.0:canais-n-presenciais/backoffice:5.7.0], middleware/acm-mediator-service-parent:[1.0.1:canais-n-presenciais/backoffice:5.7.0], middleware/backoffice-ws-parent:[2.13.0-SNAPSHOT:middleware/irc-ws-bb-parent:4.0.0, 2.17.1:canais-n-presenciais/backoffice:5.7.0], middleware/server-parent:[3.20.4:canais-n-presenciais/backoffice:5.7.0, 3.20.0-SNAPSHOT:middleware/irc-ws-bb-parent:4.0.0], canais-n-presenciais/filemanager:[2.0.0:], middleware/oauth2-manager-service-parent:[2.1.0:canais-n-presenciais/filemanager:2.0.0, 2.2.0-SNAPSHOT:middleware/irc-ws-bb-parent:4.0.0], middleware/reporting-service-parent:[2.4.0-SNAPSHOT:middleware/irc-ws-bb-parent:4.0.0, 2.4.1:canais-n-presenciais/filemanager:2.0.0], middleware/data-manager-pfs-parent:[2.10.0-SNAPSHOT:middleware/irc-ws-bb-parent:4.0.0, 2.10.2:canais-n-presenciais/filemanager:2.0.0], middleware/file-manager-service-parent:[2.7.0-SNAPSHOT:middleware/irc-ws-bb-parent:4.0.0, 2.7.0:canais-n-presenciais/filemanager:2.0.0], middleware/filemanager-jasper-reports:[1.2.0:canais-n-presenciais/filemanager:2.0.0], middleware/airtime-service-payments-parent:[2.4.0-SNAPSHOT:middleware/irc-ws-bb-parent:4.0.0, 2.3.2:], middleware/ebankaplus-jasper-reports:[1.1.0:], middleware/irc-custom-bcga-ao-parent:[2.1.1:], middleware/timwe-mediator-parent:[2.1.0:, 2.2.0-SNAPSHOT:middleware/irc-ws-bb-parent:4.0.0], middleware/unitel-mediator-parent:[2.2.0-SNAPSHOT:middleware/irc-ws-bb-parent:4.0.0, 2.1.1:], middleware/irc-ws-custom-parent:[2.1.0-SNAPSHOT:middleware/irc-ws-bb-parent:4.0.0, 2.2.1:], middleware/sms-express-mediator-parent:[3.2.0-SNAPSHOT:middleware/irc-ws-bb-parent:4.0.0, 3.1.0:], middleware/psd2-integration-layer-parent:[3.1.0-SNAPSHOT:middleware/irc-ws-bb-parent:4.0.0], middleware/smsc-mediator-parent:[2.4.0-SNAPSHOT:middleware/irc-ws-bb-parent:4.0.0, 2.3.1:], middleware/cpi-invoice-service-parent:[1.0.1:]]`

Can you check this please?

**Comment by crisgoncalves** [comment]
Q: For the `validate` command:
1. When we're already on the project's root directory, can we use `einstein validate .`, indicating that the file it's in the current working directory?
2. When we're not in the project's root directory and we need to specify the `einstein.yaml` file path, can we exclude the file name itself from this provided path? The einstein file name it's always the same so it would be useful not have to mention it everytime we run this command

File: src/main/groovy/io/github/asseco/pst/commands/ValidateCommand.groovy [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    void setInputFile(String aInputFile) {
+
+        FileValidator fileValidator = new FileValidator()
+                .setFilePath(aInputFile)
+                .setFileExtension("yaml")
+                .setLogger(logger)
+
+        if (!fileValidator.validate()) {
+            logger.error("The provided file path \"${aInputFile}\" or extension is not valid!")
+            System.exit(301)
```
Comment (crisgoncalves) [inline src/main/groovy/io/github/asseco/pst/commands/ValidateCommand.groovy:28]: I believe this an HTTP code? 
If it is, by convention, we should adopt the common codes '0' (successful) and '1' (fail) to indicate the operation status, always describing the failure reason.
Optionally, we can create custom codes and add them to a specific einstein command so the user can check the meaning of each of them?

File: src/main/groovy/io/github/asseco/pst/infrastructure/utils/SemanticVersion.groovy [hunk 2]
```diff
@@ -44,11 +41,22 @@ class SemanticVersion extends Semver {
...
 
-    static String getBiggestVersion(Set<SemanticVersion> aVersions) {
+    /**
+     * Checks if the provided version is not a range
+     *
+     * @param aVersion
+     * @return true if is not a range
+     */
+    synchronized static boolean isDeclaredVersion(String aVersion) {
+        return !Pattern.compile("[\\^x*~<>=]").matcher(aVersion).find()
```
Comment (crisgoncalves) [inline src/main/groovy/io/github/asseco/pst/infrastructure/utils/SemanticVersion.groovy:51]: Missing the '-' character used to specify ranges

**Review by crisgoncalves — CHANGES_REQUESTED** [review]
@hgpestana please check the open discussions.
Thanks

File: src/main/groovy/io/github/asseco/pst/commands/ValidateCommand.groovy (hunk 1 above)
Comment (hgpestana) [reply src/main/groovy/io/github/asseco/pst/commands/ValidateCommand.groovy:28]: @crisgoncalves actually it is supposed to be the latter. The idea is to create custom codes that, in the future, can make it easier for the user to understand why that result was returned. The fact that it is similar to an HTTP code is pure coincidence.

File: src/main/groovy/io/github/asseco/pst/infrastructure/utils/SemanticVersion.groovy (hunk 2 above)
Comment (hgpestana) [reply src/main/groovy/io/github/asseco/pst/infrastructure/utils/SemanticVersion.groovy]: Resolved in b41d8f7

**Pull request merged**